pandas/core/indexes/base.py

```diff
@@ -1959,537 +2416,432 @@ def get_duplicates(self):
-        strings ``''`` or :attr:`numpy.inf` are not considered NA values
-        (unless you set ``pandas.options.mode.use_inf_as_na = True``).
-        NA values, such as None or :attr:`numpy.NaN`, get mapped to ``False``
-        values.
+        This function adds lists and Series as valid boolean indexers
+        (ndarrays only supports ndarray with dtype=bool).
 
-        .. versionadded:: 0.20.0
+        If resulting ndim != 1, plain ndarray is returned instead of
+        corresponding `Index` subclass.
 
-        Returns
-        -------
-        numpy.ndarray
-            Boolean array to indicate which entries are not NA.
+        """
+        # There's no custom logic to be implemented in __getslice__, so it's
+        # not overloaded intentionally.
+        getitem = self._data.__getitem__
+        promote = self._shallow_copy
 
-        See Also
-        --------
-        Index.notnull : Alias of notna.
-        Index.isna: Inverse of notna.
-        pandas.notna : Top-level notna.
+        if is_scalar(key):
+            key = com.cast_scalar_indexer(key)
+            return getitem(key)
 
-        Examples
-        --------
-        Show which entries in an Index are not NA. The result is an
-        array.
+        if isinstance(key, slice):
+            # This case is separated from the conditional above to avoid
+            # pessimization of basic indexing.
+            return promote(getitem(key))
 
-        >>> idx = pd.Index([5.2, 6.0, np.NaN])
-        >>> idx
-        Float64Index([5.2, 6.0, nan], dtype='float64')
-        >>> idx.notna()
-        array([ True,  True, False])
+        if com.is_bool_indexer(key):
+            key = np.asarray(key, dtype=bool)
 
-        Empty strings are not considered NA values. None is considered a NA
-        value.
+        key = com.values_from_object(key)
+        result = getitem(key)
+        if not is_scalar(result):
+            return promote(result)
+        else:
+            return result
 
-        >>> idx = pd.Index(['black', '', 'red', None])
-        >>> idx
-        Index(['black', '', 'red', None], dtype='object')
-        >>> idx.notna()
-        array([ True,  True,  True, False])
+    def _can_hold_identifiers_and_holds_name(self, name):
         """
-        return ~self.isna()
-    notnull = notna
+        Faster check for ``name in self`` when we know `name` is a Python
+        identifier (e.g. in NDFrame.__getattr__, which hits this to support
+        . key lookup). For indexes that can't hold identifiers (everything
+        but object & categorical) we just return False.
 
-    def putmask(self, mask, value):
+        https://github.com/pandas-dev/pandas/issues/19764
         """
-        Return a new Index of the values set with the mask.
+        if self.is_object() or self.is_categorical():
+            return name in self
+        return False
 
-        See Also
-        --------
-        numpy.ndarray.putmask
```

Comment:
reshaping methods

Reply:
what method was this comment supposed to refer to?

Reply:
append/concat section

Reply:
Same, asking what method, not what section